workers/workers_test.go

```diff
@@ -131,3 +136,147 @@ func BenchmarkSerial(b *testing.B) {
 		})
 	}
 }
+
+func TestJobWait(t *testing.T) {
+	require := require.New(t)
+	job := Job{
+		tasks:     make(chan func() error),
+		completed: make(chan struct{}),
+		result:    make(chan error),
+	}
+	testError := errors.New("TestError")
+	go func() {
+		job.result <- testError
+	}()
+	returned := job.Wait()
+	require.ErrorIs(testError, returned, "Incorrect error returned.")
+}
+
+func TestJobDone(t *testing.T) {
+	require := require.New(t)
+	job := Job{
+		tasks:     make(chan func() error),
+		completed: make(chan struct{}),
+		result:    make(chan error),
+	}
+
+	called := false
+	job.Done(func() {
+		called = true
+	})
+	job.completed <- struct{}{}
+	// Check that the callback function was called
+	require.True(called, "Callback function not called")
+	// Ensure task channel was closed
+	require.Empty(job.tasks, "Tasks channel not closed")
+}
+
+func TestJobGo(t *testing.T) {
+	require := require.New(t)
+	job := Job{
+		tasks:     make(chan func() error, 2),
+		completed: make(chan struct{}),
+		result:    make(chan error),
+	}
+	n := 1
+	testError := errors.New("TestError")
+	for n <= 2 {
+		job.Go(func() error {
+			if n == 1 {
+				return nil
+			} else {
+				return testError
+			}
+		})
+		n += 1
+	}
+	n = 1
+	// Ensure corret errors
+	for n <= 2 {
+		response := <-job.tasks
+		if n == 1 {
+			require.NoError(response(), "Incorrect error message.")
+		} else {
+			require.ErrorIs(testError, response(), "Incorrect error message.")
+		}
+		n += 1
+	}
+}
+
+func TestStopWorker(t *testing.T) {
+	require := require.New(t)
+	w := New(10, 100)
+	require.False(w.shouldShutdown, "Shutdown val incorrectly initialized.")
+	w.Stop()
+	require.Empty(w.queue, "Worker queue not empty")
+	require.True(w.shouldShutdown, "Shutdown val not set.")
+}
+
+func TestWorker(t *testing.T) {
+	require := require.New(t)
+	w := New(10, 10000)
+	// Require workers was created properly
+	require.Equal(10, w.count, "Count not set correctly")
+	require.Empty(w.queue, "Worker queue not empty")
+	require.Empty(w.tasks, "Worker tasks not empty")
+	// Shutdown fields
+	require.Empty(w.ackShutdown, "Worker ackShutdown not empty")
+	require.Empty(w.stopWorkers, "Worker stopWorkers not empty")
+	require.Empty(w.stoppedWorkers, "Worker stoppedWorkers not empty")
+
+	vals := int64(0)
+	for i := 0; i < 1000; i++ {
+		// Create a new job
+		job, err := w.NewJob(10)
+		require.NoError(err)
+		for j := 0; j < 10; j++ {
+			task := func() error {
+				atomic.AddInt64(&vals, 1)
+				return nil
+			}
+			job.Go(task)
+		}
+		// TODO: Better way to do this?
+		time.Sleep(time.Millisecond * 2)
```

Comment:
not sure why leaving this out fails the test

Reply:
What is the fail that we see?

Reply:
That is odd

Reply:
May be indicative of some race between `Go` and `Done`.

Reply:
ooooo I think I know why.

You need to call `job.Wait()` in another loop somewhere. Just calling `Done()` does not block for things to finish.

Reply:
Calling `w.Stop()` before all jobs have been waited on means that you could exit before processing everything.